ivm/src/ext.rs

```diff
@@ -5,13 +5,225 @@
+      )
+    }
+    define_ext_fns!(ext_fn_map, self,
+      "seq" => |a, _b| a,
+
+      "add" => |a, b| numeric_op(n32_ext_ty, f32_ext_ty, a, b, u32::wrapping_add, f32::add),
+      "sub" => |a, b| numeric_op(n32_ext_ty, f32_ext_ty, a, b, u32::wrapping_sub, f32::sub),
+      "mul" => |a, b| numeric_op(n32_ext_ty, f32_ext_ty, a, b, u32::wrapping_mul, f32::mul),
+      "div" => |a, b| numeric_op(n32_ext_ty, f32_ext_ty, a, b, u32::wrapping_div, f32::div),
+      "rem" => |a, b| numeric_op(n32_ext_ty, f32_ext_ty, a, b, u32::wrapping_rem, f32::rem_euclid),
+
+      "eq" => |a, b| comparison(n32_ext_ty, f32_ext_ty, a, b, |a, b| a == b, |a, b| a == b),
+      "ne" => |a, b| comparison(n32_ext_ty, f32_ext_ty, a, b, |a, b| a != b, |a, b| a != b),
+      "lt" => |a, b| comparison(n32_ext_ty, f32_ext_ty, a, b, |a, b| a < b, |a, b| a < b),
+      "le" => |a, b| comparison(n32_ext_ty, f32_ext_ty, a, b, |a, b| a <= b, |a, b| a <= b),
+
+      "n32_shl" => |a, b| ExtVal::new(n32_ext_ty, a.as_ty(&n32_ext_ty).wrapping_shl(b.as_ty(&n32_ext_ty))),
+      "n32_shr" => |a, b| ExtVal::new(n32_ext_ty, a.as_ty(&n32_ext_ty).wrapping_shr(b.as_ty(&n32_ext_ty))),
+      "n32_rotl" => |a, b| ExtVal::new(n32_ext_ty, a.as_ty(&n32_ext_ty).rotate_left(b.as_ty(&n32_ext_ty))),
+      "n32_rotr" => |a, b| ExtVal::new(n32_ext_ty, a.as_ty(&n32_ext_ty).rotate_right(b.as_ty(&n32_ext_ty))),
+
+      "n32_and" => |a, b| ExtVal::new(n32_ext_ty, a.as_ty(&n32_ext_ty) & b.as_ty(&n32_ext_ty)),
+      "n32_or" => |a, b| ExtVal::new(n32_ext_ty, a.as_ty(&n32_ext_ty) | b.as_ty(&n32_ext_ty)),
+      "n32_xor" => |a, b| ExtVal::new(n32_ext_ty, a.as_ty(&n32_ext_ty) ^ b.as_ty(&n32_ext_ty)),
+
+      "n32_add_high" => |a, b| ExtVal::new(n32_ext_ty, (((a.as_ty(&n32_ext_ty) as u64) + (b.as_ty(&n32_ext_ty) as u64)) >> 32) as u32),
+      "n32_mul_high" => |a, b| ExtVal::new(n32_ext_ty, (((a.as_ty(&n32_ext_ty) as u64) * (b.as_ty(&n32_ext_ty) as u64)) >> 32) as u32),
+
+      "io_print_char" => |a, b| {
+        a.as_ty(&io_ext_ty);
+        print!("{}", char::try_from(b.as_ty(&n32_ext_ty)).unwrap());
+        ExtVal::new(io_ext_ty, 0)
+      },
+      "io_print_byte" => |a, b| {
+        a.as_ty(&io_ext_ty);
+        io::stdout().write_all(&[b.as_ty(&n32_ext_ty) as u8]).unwrap();
+        ExtVal::new(io_ext_ty, 0)
+      },
+      "io_flush" => |a, _b| {
+        a.as_ty(&io_ext_ty);
+        io::stdout().flush().unwrap();
+        ExtVal::new(io_ext_ty, 0)
+      },
+      "io_read_byte" => |a, b| {
+        a.as_ty(&io_ext_ty);
+        let default = b.as_ty(&n32_ext_ty) as u8;
+        let mut buf = [default];
+        _ = io::stdin().read(&mut buf).unwrap();
+        ExtVal::new(n32_ext_ty, buf[0] as u32)
+      }
+    );
+    (ext_fn_map, ext_ty_map)
+  }
+  pub fn call(
+    &self,
+    ext_fn: ExtFn<'ivm>,
+    mut arg0: ExtVal<'ivm>,
+    mut arg1: ExtVal<'ivm>,
+  ) -> ExtVal<'ivm> {
+    let closure =
+      self.ext_fns.get(ext_fn.kind() as usize).expect("ext_fn with given ID not found!");
```

Comment:
in theory, if the `ExtFn` is invariant on the `'ivm`, and if we ensure there can only be one `Extrinsics` for a given `'ivm` (not sure if this is upheld), this check can be removed... probably premature optimization, though; doesn't seem to impact performance

Reply:
I don't think the fact that there is only one `Extrinsics` for a given `'ivm` is upheld by the typesystem right now.